yt_dlp/extractor/peloton.py

```diff
@@ -0,0 +1,336 @@
+                else:
+                    raise ExtractorError(res['message'])
+            else:
+                raise
+        return subscription['token']
+
+    def _extract(self, video_id):
+        try:
+            self._start_session(video_id)
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 401:
+                self._login(video_id)
+                self._start_session(video_id)
+            else:
+                raise
+
+        metadata = self._get_ride(video_id)
+        if not metadata.get('ride'):
+            raise ExtractorError('Missing stream metadata')
+
+        token = self._get_token(video_id)
+
+        is_live = False
+        subtitles = {}
+        content_format = try_get(metadata, lambda x: x['ride']['content_format'], compat_str) or 'video'
+        if content_format == 'audio':
+            url = self._MANIFEST_URL_TEMPLATE % (metadata['ride'].get('vod_stream_url'), compat_urllib_parse.quote(token))
+            formats = [{
+                'url': url,
+                'ext': 'm4a',
+                'format_id': 'audio',
+                'resolution': 'audio only'
+            }]
+            ext = 'm4a'
+        else:
+            if metadata['ride'].get('vod_stream_url'):
+                url = self._PROXY_TEMPLATE % (
+                    ','.join([re.sub('^([a-z]+)-([A-Z]+)$', r'\1', caption) for caption in metadata['ride']['captions']]),
+                    self._MANIFEST_URL_TEMPLATE % (
+                        metadata['ride']['vod_stream_url'],
+                        compat_urllib_parse.quote(compat_urllib_parse.quote(token))))
+            elif metadata['ride'].get('live_stream_url'):
+                url = self._MANIFEST_URL_TEMPLATE % (metadata['ride'].get('live_stream_url'), compat_urllib_parse.quote(token))
+                is_live = True
+            else:
+                raise ExtractorError('Missing video URL')
+            formats = self._extract_m3u8_formats(url, video_id, 'mp4')
+            if formats:
+                for i, f in enumerate(formats, start=2):
+                    if f.get('vcodec') == 'none':
+                        f['source_preference'] = -i
+                    else:
+                        break
+                self._sort_formats(formats)
+            ext = 'mp4'
+            m3u8_doc = self._download_webpage(url, video_id, note='Downloading m3u8', fatal=False)
+            if m3u8_doc:
+                for line in m3u8_doc.splitlines():
+                    if line.startswith('#EXT-X-MEDIA:'):
+                        media = parse_m3u8_attributes(line)
+                        if media.get('TYPE') == 'SUBTITLES':
+                            subtitles.setdefault(media.get('LANGUAGE'), []).append({
+                                'url': media.get('URI'),
+                                'ext': 'vtt'
+                            })
```

Comment:
It would be nice if this can be generalized and made available in `common.py`. There are other extractors that can probably benefit from this